Hardcode lifetime categories

mgehre opened the pull request:
- Make DerefType optional in attributes
- Add __is_gsl_owner/__is_gsl_pointer builtins for tests
- Hard code gsl::Owner/gsl::Pointer for std types

TODO: document the builtins and that the DerefType is optional

mgehre committed:
gsl::Owner/gsl::Pointer: Make DerefType parameter optional

mgehre committed:
Add __is_gsl_owner/__is_gsl_pointer builtins

mgehre committed:
Hard code gsl::Owner/gsl::Pointer for std types

mgehre commented:
TODO: Add attributes to classes that are not directly in std::, like std::vector<>::iterator et al.

Xazax-hun commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -6036,6 +6036,57 @@ static bool canPassInRegisters(Sema &S, CXXRecordDecl *D,
+/// \pre Record->getIdentifiers() is non-null
+void Sema::addOwnerPointerAttribute(CXXRecordDecl *Record) {
+
+  static llvm::StringSet<> StdOwners{
```
For some reason LLVM folks do not really like static initializers. I windows if we could make this a const member of Sema and init it in the ctor. We could also wait with this change until the reviewers actually want it.

mgehre replied:
Note that this is not a static initializer in the sense that it runs before main - it runs on first entry to the function. I would wait for the review.
I actually would love to make this a constexpr container, but there seems to be no constexpr map in std/llvm yet.

Xazax-hun commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -6036,6 +6036,57 @@ static bool canPassInRegisters(Sema &S, CXXRecordDecl *D,
+      "unordered_map",
+      "unordered_multiset",
+      "unordered_multimap",
+      "variant",
```
What would be the deref type for variant? I think we should not add it now (since we do not have the gsl::lifetime annotation yet).

mgehre replied:
As we discussed: Without doing automatic function model inference, we don't need the deref type.
Later, we add [[gsl::lifetime]] to the `std::get()` function to make that work.
The lifetime paper explicitly says 
```
Additionally, for convenient adoption without modifying existing standard library headers, the following well-known standard types are treated as-if annotated as Owners: stack, queue, priority_queue, optional, variant, any, and regex.
```

Xazax-hun reviewed (commented):
Thanks, this looks great!

You was quite fast!

Do you think we could start filling the pointee type too? Maybe it is not much work and in the end we might not need to make the type argument optional (yet). I feel a bit bad about making it optional until we support the [[gsl::lifetime]] attribute.

Xazax-hun commented:
> 
> 
> TODO: Add attributes to classes that are not directly in std::, like std::vector<>::iterator et al.

 The `isInStdNamespace` call should return true for those as well right? Maybe we could look up the iterator typedef in STL containers. But if we need to look member types up, adding the pointee type is not that much extra work.

Xazax-hun commented on clang/test/SemaCXX/attr-gsl-owner-pointer.cpp:
```diff
@@ -59,9 +70,21 @@ class [[gsl::Owner(int)]] AddTheSameLater{};
+static_assert(__is_gsl_owner(OwnerWithEmptyParameterList), "");
 
 class [[gsl::Pointer()]] PointerWithEmptyParameterList{};
+static_assert(__is_gsl_pointer(PointerWithEmptyParameterList), "");
```
Now it is possible to write metaprograms that make decisions based on a type being a Pointer or Owner :) Of course, people not supposed to use __ names diretly, so this is not really exposed to the user. I wonder what will the reviewers think about this.

mgehre replied:
I copied the approach from https://reviews.llvm.org/D50119, which makes testing very clean. Seems to be not controversial there.

mgehre committed:
Attach Owner/Pointer attributes to non-complete template instantiations

mgehre committed:
Move addOwnerPointerAttribute to SemaAttr.cpp, where other functions of this kind live

mgehre committed:
Don't crash AST dump when Owner/Pointer has no Type argument

mgehre committed:
__is_gsl_owner/pointer: don't require complete type

mgehre committed:
Hard-code std:: Owner/Pointer already on incomplete types/forward declarations

mgehre commented:
I would guess  that the standard allows std::vector<>::iterator to be implemented in a different namespace (e.g. std::__internal) or as a member class of std::vector.

Do you know where we can put tests of the hard-coded annotations against real STL implementations? The tests in clang/ all need to be independent of the STL. Would https://github.com/llvm-mirror/test-suite be the right place?

Xazax-hun commented:
> 
> 
> I would guess that the standard allows std::vector<>::iterator to be implemented in a different namespace (e.g. std::__internal) or as a member class of std::vector.

I think the easiest way to get it done is to also do some processing on typedefs/usings and if it is inside a container like vector, and the name of the declaration is iterator or const_iterator than we can mark the desugared type as Pointer. We also need to handle the case if iterator is actually an inline class. 

> 
> Do you know where we can put tests of the hard-coded annotations against real STL implementations? The tests in clang/ all need to be independent of the STL. Would https://github.com/llvm-mirror/test-suite be the right place?

I think it is quite OK to just mock multiple STL interfaces in some headers.

mgehre commented:
> I think it is quite OK to just mock multiple STL interfaces in some headers.

Yeah, it's okay as a unit test, but it will not prove that our implementation actually works with the MSVC STL, libstdc++, libcxx and other STLs that have a build bot. I remember that we had quite some interesting edge case with the different STLs in our previous work.

mgehre committed:
Don't crash when pretty-printing an Attribute with optional type argument

mgehre committed:
Put gsl::Pointer on std::container::iterator

Xazax-hun reviewed (commented):
Wow, nice work!

mgehre committed:
__is_gsl_pointer: return true for pointers and references

mgehre committed:
Correctly annoatate std::string

mgehre committed:
clang/test/SemaCXX/attr-gsl-owner-pointer.cpp: Doc

mgehre committed:
Merge remote-tracking branch 'mgehre/lifetime-categories' into hardcode-lifetime-categories

mgehre committed:
clang-format

mgehre committed:
Add documentation to __is_gsl_* intrinsics

Xazax-hun commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -2769,14 +2769,16 @@ def TypeTagForDatatype : InheritableAttr {
   let Spellings = [CXX11<"gsl", "Owner">];
   let Subjects = SubjectList<[CXXRecord]>;
-  let Args = [TypeArgument<"DerefType">];
+  let Args = [TypeArgument<"DerefType", 1>];
```
Maybe
```
let Args = [TypeArgument<"DerefType", /*opt*/1>];
```
?

Xazax-hun commented on clang/lib/Sema/SemaAttr.cpp:
```diff
@@ -85,6 +85,116 @@ void Sema::AddMsStructLayoutForRecord(RecordDecl *RD) {
+      "unordered_multimap",
+  };
+
+  static llvm::StringSet<> Iterators{"iterator", "const_iterator"};
```
Should we add `reverse_iterator` and `const_reverse_iterator` as well?

Xazax-hun commented on clang/lib/Sema/SemaAttr.cpp:
```diff
@@ -85,6 +85,116 @@ void Sema::AddMsStructLayoutForRecord(RecordDecl *RD) {
+
+/// Add [[gsl::Owner]] and [[gsl::Pointer]] attributes for std:: types.
+void Sema::addDefaultOwnerPointerAttribute(CXXRecordDecl *Record) {
+  static llvm::StringSet<> StdOwners{
```
I wonder if we should reuse the containers dict from above.

mgehre replied:
They are not exactly the same. `Containers` does only contain classes that have iterators, i.e it does not contain `any`, `variant` and `basic_regex`.

Xazax-hun replied:
Yeah, but if there is a subset relationship we may be able utilize that. It would mean two lookups instead of one though.

Xazax-hun commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -2769,14 +2769,16 @@ def TypeTagForDatatype : InheritableAttr {
   let Subjects = SubjectList<[CXXRecord]>;
-  let Args = [TypeArgument<"DerefType">];
+  let Args = [TypeArgument<"DerefType", 1>];
+    let MeaningfulToClassTemplateDefinition = 1;
```
Indentation is off here.

Xazax-hun commented on clang/lib/Sema/SemaAttr.cpp:
```diff
@@ -85,6 +85,116 @@ void Sema::AddMsStructLayoutForRecord(RecordDecl *RD) {
+
+    if (StdOwners.count(Record->getName())) {
+      addOwnerPointerAttributeIfNotExisting<OwnerAttr>(Context, Record);
+    } else if (StdPointers.count(Record->getName())) {
```
Redundant braces.

Xazax-hun commented on clang/lib/Sema/SemaAttr.cpp:
```diff
@@ -85,6 +85,116 @@ void Sema::AddMsStructLayoutForRecord(RecordDecl *RD) {
+/// Add gsl::Pointer attribute to std::container::iterator
+/// \param ND The declaration that introduces the name std::container::iterator.
+/// \param UnderlyingRecord The record named by ND.
+void Sema::addDefaultOwnerPointerAttribute(NamedDecl *ND,
```
In case this only ever adds Pointer attribute maybe we should get rid of the Owner from the name?

Xazax-hun commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -4550,7 +4553,10 @@ static void handleLifetimeCategoryAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
       return;
     if (const auto *OAttr = D->getAttr<OwnerAttr>()) {
-      if (OAttr->getDerefType().getTypePtr() != ParmType.getTypePtr()) {
+      const Type *ExistingDerefType = OAttr->getDerefTypeLoc()
```
Do we actually need to have the ternary operator? Could we just rely on comparing to QualTypes? Or two invalid QualTypes compared inequal? Then we could use getTypePtrOrNull on both of them.

mgehre replied:
`getDerefType` dereferences a null pointer when `getDerefTypeLoc` is null.

Xazax-hun reviewed (commented):
Great work, some nits inline :)

mgehre committed:
Implement comments

Xazax-hun commented on clang/test/SemaCXX/attr-gsl-owner-pointer.cpp:
```diff
@@ -131,7 +131,7 @@ template <
     class Traits>
 class basic_regex;
-static_assert(__is_gsl_pointer(basic_regex<char, void>), "");
+static_assert(__is_gsl_owner(basic_regex<char, void>), "");
```
Don't we have a static_assert overload that does not take a second argument?

mgehre replied:
Since C++17, yes. Is it safe to change the test to run in C++17 mode? Most tests seem to just use the default language mode. I copied that part also from https://reviews.llvm.org/D50119

Xazax-hun replied:
I see, makes sense.

mgehre committed:
clang/include/clang/Basic/AttrDocs.td: Say that DerefType is optional

mgehre committed:
gsl::Owner/gsl::Pointer: Add implicit annotations for some std types

Summary:
Make the DerefType, i.e. the argument of gsl::Owner/gsl::Pointer optional for now.

The DerefType is used when infering lifetime annotations of functions, e.g.
how a return value can relate to the arguments. Our initial checks should
work without that kind of inference and instead use explicit annotations,
so we don't need the DerefType right now.
And having it optional makes the implicit/default annotations for std:: types
easier to implement. I think the DerefType is also optional in the current MSVC
implementation.

This seems to be the first attribute that has an optional Type argument,
so I needed to fix two thinks in tablegen:
- Don't crash AST dump when Owner/Pointer has no Type argument
- Don't crash when pretty-printing an Attribute with optional type argument

Add __is_gsl_owner/__is_gsl_pointer builtins. They are very useful to test
if the annotations have been attached correctly.

Hard code gsl::Owner/gsl::Pointer for std types. The paper mentions
some types explicitly. Generally, all containers and their iterators are
covered. For iterators, we cover both the case that they are defined
as an nested class or as an typedef/using. I have started to test this
implementation against some real standard library implementations, namely
libc++ 7.1.0, libc++ 8.0.1rc2, libstdc++ 4.6.4, libstdc++ 4.8.5,
libstdc++ 4.9.4, libstdc++ 5.4.0, libstdc++ 6.5.0, libstdc++ 7.3.0,
libstdc++ 8.3.0 and libstdc++ 9.1.0.

The tests are currently here
  https://github.com/mgehre/llvm-project/blob/lifetime-ci/lifetime-attr-test.sh
  https://github.com/mgehre/llvm-project/blob/lifetime-ci/lifetime-attr-test.cpp
I think due to their dependency on a standard library, they are not a good fit
for clang/test/. Where else could I put them?

Reviewers: gribozavr, xazax.hun

Subscribers: rnkovacs, cfe-commits

Tags: #clang

Differential Revision: https://reviews.llvm.org/D64448

mgehre committed:
Implement comments

mgehre committed:
Merge branch 'lifetime-categories' into hardcode-lifetime-categories; Split test; Move test to AST dump; fix for comments;

mgehre committed:
Remove type traits

Xazax-hun commented:
As far as I understand this is not relevant anymore :)
